sysfs/class_infiniband.go

```diff
@@ -0,0 +1,372 @@
+	PortRcvConstraintErrors  *uint64 // counters/port_rcv_constraint_errors
+	PortRcvData              *uint64 // counters/port_rcv_data
+	PortRcvDiscards          *uint64 // counters/port_rcv_discards
+	PortRcvErrors            *uint64 // counters/port_rcv_errors
+	PortRcvPackets           *uint64 // counters/port_rcv_packets
+	PortXmitConstraintErrors *uint64 // counters/port_xmit_constraint_errors
+	PortXmitData             *uint64 // counters/port_xmit_data
+	PortXmitDiscards         *uint64 // counters/port_xmit_discards
+	PortXmitPackets          *uint64 // counters/port_xmit_packets
+	PortXmitWait             *uint64 // counters/port_xmit_wait
+	UnicastRcvPackets        *uint64 // counters/unicast_rcv_packets
+	UnicastXmitPackets       *uint64 // counters/unicast_xmit_packets
+}
+
+// InfiniBandPort contains info from files in
+// /sys/class/infiniband/<Name>/ports/<Port>
+// for a single port of one InfiniBand device.
+type InfiniBandPort struct {
+	Name        string
+	Port        uint
+	State       string // String representation from /sys/class/infiniband/<Name>/ports/<Port>/state
+	StateID     uint   // ID from /sys/class/infiniband/<Name>/ports/<Port>/state
+	PhysState   string // String representation from /sys/class/infiniband/<Name>/ports/<Port>/phys_state
+	PhysStateID uint   // String representation from /sys/class/infiniband/<Name>/ports/<Port>/phys_state
+	Rate        uint64 // in bytes/second from /sys/class/infiniband/<Name>/ports/<Port>/rate
+	Counters    InfiniBandCounters
+}
+
+// InfiniBandDevice contains info from files in /sys/class/infiniband for a
+// single InfiniBand device.
+type InfiniBandDevice struct {
+	Name            string
+	BoardID         string // /sys/class/infiniband/<Name>/board_id
+	FirmwareVersion string // /sys/class/infiniband/<Name>/fw_ver
+	HCAType         string // /sys/class/infiniband/<Name>/hca_type
+	Ports           map[uint]InfiniBandPort
+}
+
+// InfiniBandClass is a collection of every InfiniBand device in
+// /sys/class/infiniband.
+//
+// The map keys are the names of the InfiniBand devices.
+type InfiniBandClass map[string]InfiniBandDevice
+
+// InfiniBandClass returns info for all InfiniBand devices read from
+// /sys/class/infiniband.
+func (fs FS) InfiniBandClass() (InfiniBandClass, error) {
+	path := fs.sys.Path(infinibandClassPath)
+
+	dirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list InfiniBand devices at %q: %v", path, err)
+	}
+
+	ibc := make(InfiniBandClass, len(dirs))
+	for _, d := range dirs {
+		device, err := fs.parseInfiniBandDevice(d.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		ibc[device.Name] = *device
+	}
+
+	return ibc, nil
+}
+
+// Parse one InfiniBand device.
+func (fs FS) parseInfiniBandDevice(name string) (*InfiniBandDevice, error) {
+	path := fs.sys.Path(infinibandClassPath, name)
+	device := InfiniBandDevice{Name: name}
+
+	for _, f := range [3]string{"board_id", "fw_ver", "hca_type"} {
```

Comment:
Nit: how about `[...]` so the size will be inferred automatically? I'd usually just go with a slice but an array seems reasonable since this will probably never change.

Reply:
Done.